### PR TITLE
Adapter for Constant from opset 9 to 8

### PR DESCRIPTION
**Pull request by KsenijaS**
Add a adapter for constant from opset 9 to 8

### CODE REVIEW

**Comment by KsenijaS** [comment]
This PR fixes the issue:
https://github.com/onnx/onnx/issues/1986

**Comment by spandantiwari** [comment]
@KsenijaS - Thanks for this bug fix. Will you be able to look at this further, maybe pass the CI? Or if you like, we could help. Thanks.

**Comment by KsenijaS** [comment]
I will continue working on this

**Comment by askhade** [comment]
Your CIs are failing with error :  AttributeError: type object 'Callable' has no attribute '_abc_registry'
This was a CI pipeline issue which was fixed a few days back. Please merge with master and this will be resolved.

**Comment by KsenijaS** [comment]
PR is ready to be merged, just update it to master branch first.

File: onnx/version_converter/adapters/constant_9_8.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      post_cast->output()->setUniqueName(original_output_name);
+      post_cast->output()->setSizes(outputs[0]->sizes());
+      post_cast->output()->setElemType(output_type);
+
+      post_cast->insertAfter(node);
+
+      for (Use u : original_uses) {
+        u.user->replaceInputWith(node->output(), post_cast->output());
+      }
+
```
Comment (spandantiwari) [inline onnx/version_converter/adapters/constant_9_8.h]: nit: extra line.

File: onnx/version_converter/convert.h [hunk 2]
```diff
@@ -29,6 +29,8 @@
 #include "onnx/version_converter/adapters/scan_9_8.h"
 #include "onnx/version_converter/adapters/scan_8_9.h"
 #include "onnx/version_converter/adapters/cast_9_8.h"
+#include "onnx/version_converter/adapters/constant_9_8.h"
+
```
Comment (spandantiwari) [inline onnx/version_converter/convert.h]: nit: extra line

File: onnx/version_converter/adapters/constant_9_8.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    const ArrayRef<Value*>& outputs = node->outputs();
+    const std::string original_output_name = node->output()->uniqueName();
+    const int output_type = outputs[0]->elemType();
+    int const_type = TensorProto_DataType::TensorProto_DataType_FLOAT;
+
+    if (output_type == TensorProto_DataType::TensorProto_DataType_FLOAT ||
+        output_type == TensorProto_DataType::TensorProto_DataType_FLOAT16 ||
+        output_type == TensorProto_DataType::TensorProto_DataType_DOUBLE)
+        return;
```
Comment (spandantiwari) [inline onnx/version_converter/adapters/constant_9_8.h]: nit: can we enclose this single line (return) in braces, just to keep the style of the rest of the code.

File: onnx/version_converter/convert.h (hunk 2 above)
Comment (wschin) [inline onnx/version_converter/convert.h]:
```suggestion
```

File: onnx/version_converter/adapters/constant_9_8.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        TensorProto_DataType::TensorProto_DataType_UINT16,
+    };
+
+    const std::unordered_set<int> &cast_to_double_types = {
+        TensorProto_DataType::TensorProto_DataType_UINT32, 
+        TensorProto_DataType::TensorProto_DataType_INT64,
+        TensorProto_DataType::TensorProto_DataType_UINT64,
+    };
+
+    if (cast_to_float_types.find(output_type) != cast_to_float_types.end()) {
```
Comment (houseroad) [inline onnx/version_converter/adapters/constant_9_8.h]: We should clearly declare what the casting rules we use here at the beginning of this file.

File: onnx/test/version_converter_test.py [hunk 5]
```diff
@@ -646,6 +647,34 @@ def test_constant_9_8(self):  # type: () -> None
         assert converted_model.graph.output[0].type.tensor_type.elem_type == data_type
         assert converted_model.opset_import[0].version == to_opset
 
+    # Test Constant Adapter: 9 -> 8
+    def test_constant_9_8_polish(self):  # type: () -> None
+        from_opset = 9
+        to_opset = 8
+        data_type = TensorProto.UINT64
```
Comment (houseroad) [inline onnx/test/version_converter_test.py]: Can we also add tests for UINT64, and FLOAT/DOUBLE cases

Comment (KsenijaS) [reply onnx/test/version_converter_test.py]: We have a test for UINT64.

**Comment by jcwchen** [comment]
This should be covered in the main branch now: https://github.com/onnx/onnx/blob/a92870cdf359297495a118184dca2eaecee4b717/onnx/version_converter/convert.h#L265. Close this PR. Thanks for the contribution!

**Pull request closed without merging**